pkg/library/kubernetes/util.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2019-2022 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"fmt"
+
+	dw "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/devworkspace-operator/pkg/common"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func HasKubelikeComponent(workspace *common.DevWorkspaceWithConfig) bool {
+	for _, component := range workspace.Spec.Template.Components {
+		if component.Kubernetes != nil || component.Openshift != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func filterForKubeLikeComponents(components []dw.Component) ([]dw.Component, error) {
+	var k8sLikeComponents []dw.Component
+	for _, component := range components {
+		k8sLikeComponent, err := getK8sLikeComponent(component)
+		if err != nil {
+			continue
+		}
+		if k8sLikeComponent.Uri != "" {
+			return nil, &FailError{fmt.Errorf("kubernetes/openshift components that define a URI are unsupported (component %s)", component.Name)}
+		}
+		if k8sLikeComponent.Inlined == "" {
+			continue
+		}
+		if k8sLikeComponent.GetDeployByDefault() {
```

Comment:
design question, my undestanding that for k8s / openshift components user explicitly need to set `deployByDefault=true`? 
As a developer, I'm wondering about the use-case when I craft a devfile, add the k8s component, and do not want it to be applied. For me it is not really clear the necessity of the `DeployByDefault` field in general

Reply:
IIRC the idea behind `deployByDefault = false` is that an `apply` command could be used to create the resource on demand -- e.g. I run a VSCode task that creates a service/deployment on the cluster. I believe it's more relevant to the `odo` use case than it is for DWO, since in our case you could just store the yaml to apply in the repo itself. The original devfile/api issue is https://github.com/devfile/api/issues/693.

One potential use case could be for creating resources without worrying about DWO's serviceAccount having the correct permissions.